src/npz.rs

```diff
@@ -243,3 +253,545 @@ impl<R: Read + Seek> NpzReader<R> {
         Ok(ArrayBase::<S, D>::read_npy(self.zip.by_index(index)?)?)
     }
 }
+
+/// An error viewing a `.npz` file.
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum ViewNpzError {
+    /// An error caused by the zip archive.
+    Zip(ZipError),
+    /// An error caused by viewing an inner `.npy` file.
+    Npy(ViewNpyError),
+    /// The start of the data is not properly aligned for the element type.
+    MisalignedData,
+    /// A mutable `.npy` file view has already been moved out of its `.npz` file view.
+    MovedNpyViewMut,
+}
+
+impl Error for ViewNpzError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            ViewNpzError::Zip(err) => Some(err),
+            ViewNpzError::Npy(err) => Some(err),
+            ViewNpzError::MisalignedData => None,
+            ViewNpzError::MovedNpyViewMut => None,
+        }
+    }
+}
+
+impl fmt::Display for ViewNpzError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ViewNpzError::Zip(err) => write!(f, "zip file error: {}", err),
+            ViewNpzError::Npy(err) => write!(f, "error viewing npy file in npz archive: {}", err),
+            ViewNpzError::MisalignedData => write!(
+                f,
+                "cannot cast unaligned array data, try 'rezip in.npz -o out.npz'"
+            ),
+            ViewNpzError::MovedNpyViewMut => write!(
+                f,
+                "mutable npy file view already moved out of npz file view"
+            ),
+        }
+    }
+}
+
+impl From<ZipError> for ViewNpzError {
+    fn from(err: ZipError) -> ViewNpzError {
+        ViewNpzError::Zip(err)
+    }
+}
+
+impl From<ViewNpyError> for ViewNpzError {
+    fn from(err: ViewNpyError) -> ViewNpzError {
+        match err {
+            ViewNpyError::MisalignedData => ViewNpzError::MisalignedData,
+            _ => ViewNpzError::Npy(err),
+        }
+    }
+}
+
+/// Immutable view for memory-mapped `.npz` files.
+///
+/// The primary use-case for this is viewing `.npy` files within a memory-mapped
+/// `.npz` archive.
+///
+/// # Notes
+///
+/// - For types for which not all bit patterns are valid, such as `bool`, the
+///   implementation iterates over all of the elements when creating the view
+///   to ensure they have a valid bit pattern.
+///
+/// - The `.npy` files within the `.npz` archive must be at least 16-byte or
```

Comment:
Right, it must just be properly aligned for the element type. Will update the docs.

Reply:
But the archive itself must also be properly aligned to meet the maximum alignment requirement of its arrays. That would be at least 8-byte to guarantee proper alignment of any supported element type of this crate for now. NumPy supports `complex128` and I guess that's why the first header version guarantees 16-byte alignment. The ideal alignment is 64-byte to enable AVX-512 which is now the default for NumPy. What about:

```text
 - The data in the buffer containing an `.npz` archive must be properly
   aligned for its `.npy` file with the maximum alignment requirement for its
   element type. Typically, this should not be a concern for memory-mapped
   files (unless an option like `MAP_FIXED` is used), since memory mappings
   are usually aligned to a page boundary.
 - The `.npy` files within the `.npz` archive must be properly aligned for
   their element type. Archives not created by this crate can be aligned with
   the help of the CLI tool [`rezip`] as in `rezip in.npz -o out.npz` which by
   default aligns to 64-byte covering all element types and additionally
   suffices larger alignment requirements for SIMD extensions up to AVX-512.
```